### DISCUSSION TITLE
Warning: Prop `className` did not match.

### DISCUSSION THREAD

**Discussion by brunouber** [vercel/next.js#50162] [Help]
### Summary

Hope this is the correct place to post this content, that is something in between a question, and a bug report.

I've a new next.js project, that uses the `app/` directory.
For styling we use styled-components (I followed the setup instructions on [next.js doc site](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-components)).
The reason why we used styled-components (even if the project is new) is that we already have a library of reusable components we would like to build on.

Everything seems to work fine, but in the browser console, this warning appears:

```
Warning: Prop `className` did not match. Server: "a1 b c1 d" Client: "a2 b c2 d"
```

**The question**: I understand what this mean, but I don't get whether this is something I should be worried about; after all everything seems to work fine... can you explain what bad things could happen if I just ignore this warning?

Then, what makes everything worse, is that we don't always get the warning.
For example, at the beginning we had something as simple as:

```
// app/layout.tsx
import { Header } from "@/components/header";
import { PageWrapper } from "@/components/styles";

const RootLayout: FC<PropsWithChildren> = ({ children }) => {
  return (
    <html lang="en">
      <body className={inter.className}>
        <Providers>
          <Header />
          <PageWrapper>{children}</PageWrapper>
        </Providers>
      </body>
    </html>
  );
};

// app/page.tsx
import { Tabs } from "@/components/tabs";

const PageHome = () => {
  <main>
    <Tabs
      tabs={[
        {
          content: (
            <Suspense fallback={<div>Loading...</div>}>
              {/* @ts-expect-error Waiting for microsoft/TypeScript/pull/51328 */}
              <TableApplications />
            </Suspense>
          ),
          dataBt: "applications",
          header: "Applications",
          id: "applications",
          selected: true,
        },
      ]}
    />
  </main>
};
```

Where `Header`, and `Tabs` are two components from our private nmp registry (hence, installed as dependencies, and loaded from node_modules).
We import those from the "@/components/" folder so that we can add the `use client` directive.

With this setup the warning never appeared.
Later we added a new component, `<BoxNotifications />`, in the layout.

```
// app/layout.tsx
import { Header } from "@/components/header";
import { BoxNotifications } from "@/components/notifications";
import { PageWrapper } from "@/components/styles";

const RootLayout: FC<PropsWithChildren> = ({ children }) => {
  return (
    <html lang="en">
      <body className={inter.className}>
        <Providers>
          <Header />
          <PageWrapper>{children}</PageWrapper>
          <BoxNotifications />
        </Providers>
      </body>
    </html>
  );
};
```

And this change caused the warning to appear.
Then, if I move the <BoxNotifications /> inside the page component, the warning  is triggered depending on where exactly I place it respect the `Tabs` component.

I understand this is hard to follow especially without a simplified repro, but given the npm registry is private there's much I can do about it - hope someone who had similar issue in the past could help me though.

Thanks

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64
    Binaries:
      Node: 20.1.0
      npm: 9.6.4
      Yarn: 1.17.3
      pnpm: N/A
    Relevant packages:
      next: 13.4.3
      eslint-config-next: 13.4.3
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
```

**Comment by ghost** [comment]
I have the same problem when using a styled component from a `loading.tsx` file.

```tsx
// app/dashboard/loading.tsx
import Loading from "@/lib/Loading";

export default function LoadingPage() {
  return <Loading />;
}
```

```tsx
// lib/Loading.tsx
"use client";

import styled from "styled-components";

export default function Loading() {
  return <LoadingStyled>Loading...</LoadingStyled>;
}

const LoadingStyled = styled.div`
  font-weight: 600;
`;
```
